File: src/Rafiq.Application/Features/Admin/AdminModels.cs
namespace Rafiq.Application.Features.Admin;

public sealed record AdminDashboardDto(
    int TotalUsers,
    int ActiveUsers,
    int TotalProfiles,
    int ManagedProfiles,
    int AppointmentsToday,
    int AppointmentsThisMonth,
    int PendingAppointments,
    int CompletedAppointments,
    int MedicationRemindersToday,
    int MedicalDocuments,
    int AiConversations,
    int NewRegistrationsThisMonth,
    decimal MonthlyGrowthPercent,
    IReadOnlyList<AdminTrendPointDto> UserGrowth,
    IReadOnlyList<AdminTrendPointDto> AppointmentTrend,
    IReadOnlyList<AdminDistributionItemDto> GenderDistribution,
    IReadOnlyList<AdminRecentUserDto> RecentUsers,
    IReadOnlyList<AdminRecentAppointmentDto> RecentAppointments);

public sealed record AdminTrendPointDto(string Label, int Value);

public sealed record AdminDistributionItemDto(string Label, int Value);

public sealed record AdminRecentUserDto(
    Guid Id,
    string FullName,
    string Email,
    bool IsActive,
    DateTime CreatedAt,
    string? ProfileImageUrl);

public sealed record AdminRecentAppointmentDto(
    Guid Id,
    string Title,
    string Provider,
    string PatientName,
    DateTime AppointmentDateTime,
    string Status);

public sealed class AdminUserQuery
{
    public string? Search { get; init; }
    public string? Status { get; init; }
    public string? Role { get; init; }
    public string SortBy { get; init; } = "createdAt";
    public string SortDirection { get; init; } = "desc";
    public int Page { get; init; } = 1;
    public int PageSize { get; init; } = 20;
}

public sealed record AdminUserListItemDto(
    Guid Id,
    string FirstName,
    string LastName,
    string Email,
    string? PhoneNumber,
    string Role,
    bool IsActive,
    bool EmailConfirmed,
    bool PhoneNumberConfirmed,
    DateTime CreatedAt,
    string? ProfileImageUrl,
    bool HasHealthProfile);

public sealed record PagedResult<T>(
    IReadOnlyList<T> Items,
    int Page,
    int PageSize,
    int TotalCount)
{
    public int TotalPages => TotalCount == 0
        ? 0
        : (int)Math.Ceiling(TotalCount / (double)PageSize);
}

// ── Audit Logs ─────────────────────────────────────────────────────────────

public sealed record AuditChangeDto(string Field, string? Before, string? After);

public sealed record AuditLogDto(
    Guid Id,
    DateTime Timestamp,
    string ActorName,
    string ActorEmail,
    string Module,
    string Action,
    string Target,
    string Severity,
    string Description,
    IReadOnlyList<AuditChangeDto> Changes);

public sealed class AuditLogQuery
{
    public string? Search    { get; init; }
    public string? Module    { get; init; }
    public string? Severity  { get; init; }
    public string? DateFrom  { get; init; }
    public string? DateTo    { get; init; }
    public int Page          { get; init; } = 1;
    public int PageSize      { get; init; } = 10;
}

// ── AI Operations ──────────────────────────────────────────────────────────

// Overview

/// <summary>Per-day quality snapshot for the 7-day trend chart.</summary>
public sealed record AiQualityDayDto(
    string Date,
    int    Requests,
    int    Failed,
    int    ThumbsUp,
    int    ThumbsDown);

/// <summary>
/// Composite AI health overview. HealthScore is derived:
///   successRate7d × 0.65 + positiveRate × 0.35
/// </summary>
public sealed record AiOverviewDto(
    int     AiHealthScore,           // 0-100 composite quality signal
    decimal PositiveRatePercent,     // thumbs-up / total reactions
    int     UnreviewedNegative,      // thumbs-down with TriageStatus = New
    int     FailedLast7Days,
    int     AvgResponseTimeMs,
    int     RequestsToday,
    int     TotalChatConversations,
    int     TotalVoiceSessions,
    IReadOnlyList<AiQualityDayDto> QualityTrend);

// Conversations

public sealed class AiRequestQuery
{
    public string?   Search        { get; init; }
    public string?   Feature       { get; init; }
    public string?   Status        { get; init; } // "success" | "failed"
    public DateTime? From          { get; init; }
    public DateTime? To            { get; init; }
    public string    SortBy        { get; init; } = "createdAt";
    public string    SortDirection { get; init; } = "desc";
    public int       Page          { get; init; } = 1;
    public int       PageSize      { get; init; } = 20;
}

public sealed record AiRequestListItemDto(
    Guid     Id,
    Guid?    UserId,
    string?  UserName,
    string?  UserEmail,
    string   Feature,
    string   ModelName,
    bool     Success,
    string?  ErrorType,
    int      DurationMs,
    DateTime CreatedAt);

/// <summary>Detail payload for the conversations drawer.</summary>
public sealed record AiRequestDetailDto(
    Guid     Id,
    string   Feature,
    string   ModelName,
    bool     Success,
    string?  ErrorType,
    int      DurationMs,
    DateTime CreatedAt,
    Guid?    UserId,
    string?  UserName,
    string?  UserEmail,
    Guid?    ConversationId,
    string?  UserPrompt,   // last user message in the conversation
    string?  AiResponse);  // last assistant message in the conversation

// Feedback

public sealed class AiFeedbackQuery
{
    public string?   Search    { get; init; }
    public string?   Reaction  { get; init; } // "ThumbsUp" | "ThumbsDown"
    public string?   Status    { get; init; } // "New" | "Investigating" | "Resolved" | "Ignored"
    public string?   Category  { get; init; }
    public DateTime? DateFrom  { get; init; }
    public DateTime? DateTo    { get; init; }
    public int       Page      { get; init; } = 1;
    public int       PageSize  { get; init; } = 20;
}

/// <summary>AI quality insights derived from the last 30 days of telemetry and feedback.</summary>
public sealed record AiInsightsDto(
    string? MostUsedFeature,
    int     MostUsedFeatureCount,
    string? HighestErrorFeature,
    int     HighestErrorCount,
    string? SlowestFeature,
    int     SlowestFeatureAvgMs,
    string? FastestFeature,
    int     FastestFeatureAvgMs,
    string? MostReportedCategory,
    int     MostReportedCategoryCount);

public sealed record AiFeedbackListItemDto(
    Guid     Id,
    Guid     UserId,
    string   UserName,
    string   UserEmail,
    string   Reaction,       // "ThumbsUp" | "ThumbsDown"
    string?  UserComment,    // free-text comment the user wrote
    string?  Category,       // predefined category string
    string   TriageStatus,   // "New" | "Investigating" | "Resolved" | "Ignored"
    string?  AdminNotes,
    string   MessageExcerpt, // first ~80 chars — compact row preview
    string   MessageContent, // full AI message content — shown when expanded
    string?  UserPrompt,     // the user question that preceded this AI response
    Guid     ConversationId,
    DateTime CreatedAt);

public sealed record UpdateAiFeedbackDto(
    string  TriageStatus,
    string? Category,
    string? AdminNotes);

// Performance

public sealed record AiDailyStatDto(
    string Date,
    int    Requests,
    int    Failed,
    int    AvgDurationMs);

public sealed record AiPerformanceDto(
    IReadOnlyList<AiDailyStatDto>           Daily,
    IReadOnlyList<AdminDistributionItemDto> RequestsByFeature,
    int     SlowRequests,
    int     P95LatencyMs,
    int     AvgLatencyMs,
    decimal ErrorRatePercent);

// ── Usage Intelligence ──────────────────────────────────────────────────────

public sealed record UsageIntelligenceOverviewDto(
    int TotalAiRequests,
    int FlaggedRequests,
    int UsersNeedingReview,
    int WarningsSent);

public sealed record UsageAttentionUserDto(
    Guid     UserId,
    string   UserName,
    string   UserEmail,
    string?  ProfileImageUrl,
    int      TotalRequests,
    int      FlaggedRequests,
    int      WarningsSent,
    DateTime? LastActivity);

public sealed record UsageFlaggedRequestDto(
    Guid     Id,
    string   RequestType,
    string   UserRequest,
    string   AiResponse,
    string   Classification,
    string   Reason,
    DateTime CreatedAt);

public sealed record UsageAdminActionDto(
    Guid     Id,
    string   ActionType,
    string   AdminName,
    string?  Notes,
    DateTime CreatedAt);

public sealed record UsageUserDetailDto(
    Guid                              UserId,
    string                            UserName,
    string                            UserEmail,
    string?                           ProfileImageUrl,
    int                               TotalRequests,
    int                               FlaggedRequests,
    int                               WarningsSent,
    DateTime?                         LastActivity,
    bool                              IsAiRestricted,
    bool                              IsRestricted,
    bool                              IsSuspended,
    IReadOnlyList<UsageFlaggedRequestDto> FlaggedItems,
    IReadOnlyList<UsageAdminActionDto>    ActionHistory);

public sealed class UsageAttentionQueueQuery
{
    public int Page     { get; init; } = 1;
    public int PageSize { get; init; } = 20;
}

public sealed record TakeUsageActionDto(
    string  ActionType,
    string? Notes);

File: RafiqAngular/src/app/Features/admin/models/admin.models.ts
export interface AdminTrendPoint {
  label: string;
  value: number;
}

export interface AdminDistributionItem {
  label: string;
  value: number;
}

export interface AdminRecentUser {
  id: string;
  fullName: string;
  email: string;
  isActive: boolean;
  createdAt: string;
  profileImageUrl?: string | null;
}

export interface AdminRecentAppointment {
  id: string;
  title: string;
  provider: string;
  patientName: string;
  appointmentDateTime: string;
  status: string;
}

export interface AdminDashboard {
  totalUsers: number;
  activeUsers: number;
  totalProfiles: number;
  managedProfiles: number;
  medicationRemindersToday: number;
  medicalDocuments: number;
  aiConversations: number;
  totalAiRequests: number;
  flaggedAiRequests: number;
  mostActiveAiUser: string | null;
  newRegistrationsThisMonth: number;
  monthlyGrowthPercent: number;
  userGrowth: AdminTrendPoint[];
  genderDistribution: AdminDistributionItem[];
  recentUsers: AdminRecentUser[];
  usersNeedAttention: AdminDashboardAttentionUser[];
}

export interface AdminDashboardAttentionUser {
  userId: string;
  userName: string;
  profileImageUrl?: string | null;
  reason: string;
}

export interface AdminUser {
  id: string;
  firstName: string;
  lastName: string;
  email: string;
  phoneNumber?: string | null;
  role: string;
  isActive: boolean;
  emailConfirmed: boolean;
  phoneNumberConfirmed: boolean;
  createdAt: string;
  profileImageUrl?: string | null;
  hasHealthProfile: boolean;
}

export interface PagedResult<T> {
  items: T[];
  page: number;
  pageSize: number;
  totalCount: number;
  totalPages: number;
}

export interface AdminUserQuery {
  search?: string;
  status?: 'active' | 'inactive' | '';
  role?: 'Admin' | 'User' | '';
  sortBy?: 'createdAt' | 'name' | 'email' | 'status';
  sortDirection?: 'asc' | 'desc';
  page?: number;
  pageSize?: number;
}

// ── AI Operations ──────────────────────────────────────────────────────────

/** Per-day entry for the 7-day quality trend chart. */
export interface AiQualityDay {
  date: string;
  requests: number;
  failed: number;
  thumbsUp: number;
  thumbsDown: number;
}

/** Composite quality overview — health score + quality KPIs. */
export interface AiOverview {
  aiHealthScore: number;           // 0-100 composite metric
  positiveRatePercent: number;     // thumbs-up / total reactions
  unreviewedNegative: number;      // unreviewed negative feedback count
  failedLast7Days: number;
  avgResponseTimeMs: number;
  requestsToday: number;
  totalChatConversations: number;
  totalVoiceSessions: number;
  qualityTrend: AiQualityDay[];    // last 7 days
}

// Conversations

export interface AiRequestItem {
  id: string;
  userId?: string | null;
  userName?: string | null;
  userEmail?: string | null;
  feature: string;
  modelName: string;
  success: boolean;
  errorType?: string | null;
  durationMs: number;
  createdAt: string;
}

export interface AiRequestQuery {
  search?: string;
  feature?: string;
  status?: 'success' | 'failed' | '';
  from?: string;
  to?: string;
  sortBy?: 'createdAt' | 'duration' | 'feature';
  sortDirection?: 'asc' | 'desc';
  page?: number;
  pageSize?: number;
}

/** Detail payload shown in the conversations drawer. */
export interface AiRequestDetail {
  id: string;
  feature: string;
  modelName: string;
  success: boolean;
  errorType?: string | null;
  durationMs: number;
  createdAt: string;
  userId?: string | null;
  userName?: string | null;
  userEmail?: string | null;
  conversationId?: string | null;
  userPrompt?: string | null;
  aiResponse?: string | null;
}

// Feedback

export interface AiFeedbackItem {
  id: string;
  userId: string;
  userName: string;
  userEmail: string;
  reaction: string;           // 'ThumbsUp' | 'ThumbsDown'
  userComment?: string | null;
  category?: string | null;
  triageStatus: string;       // 'New' | 'Investigating' | 'Resolved' | 'Ignored'
  adminNotes?: string | null;
  messageExcerpt: string;     // ≤80 chars — shown in compact row
  messageContent: string;     // full AI message — shown when expanded
  userPrompt?: string | null; // user question that preceded this AI response
  conversationId: string;
  createdAt: string;
}

export interface AiInsights {
  mostUsedFeature?: string | null;
  mostUsedFeatureCount: number;
  highestErrorFeature?: string | null;
  highestErrorCount: number;
  slowestFeature?: string | null;
  slowestFeatureAvgMs: number;
  fastestFeature?: string | null;
  fastestFeatureAvgMs: number;
  mostReportedCategory?: string | null;
  mostReportedCategoryCount: number;
}

export interface AiFeedbackQuery {
  search?: string;
  reaction?: 'ThumbsUp' | 'ThumbsDown' | '';
  status?: string;
  category?: string;
  dateFrom?: string;
  dateTo?: string;
  page?: number;
  pageSize?: number;
}

export interface UpdateAiFeedbackRequest {
  triageStatus: string;
  category?: string | null;
  adminNotes?: string | null;
}

// Performance

export interface AiDailyStat {
  date: string;
  requests: number;
  failed: number;
  avgDurationMs: number;
}

export interface AiPerformance {
  daily: AiDailyStat[];
  requestsByFeature: AdminDistributionItem[];
  slowRequests: number;
  p95LatencyMs: number;
  avgLatencyMs: number;
  errorRatePercent: number;
}


// ── Reviews ────────────────────────────────────────────────────────────────

export type ReviewStatus = 'Pending' | 'Reviewed' | 'Resolved' | 'Archived';
export type ReviewCategory = 'General' | 'BugReport' | 'FeatureRequest' | 'Performance' | 'UiUx' | 'ContentQuality';

export interface AdminReview {
  id: string;
  displayName: string;
  stars: number;
  comment: string | null;
  isVisible: boolean;
  status: ReviewStatus;
  category: ReviewCategory;
  adminNotes: string | null;
  adminReply: string | null;
  repliedAt: string | null;
  reviewedAt: string | null;
  deviceInfo: string | null;
  appVersion: string | null;
  appLanguage: string | null;
  createdAt: string;
}

export interface AdminReviewsPage {
  items: AdminReview[];
  total: number;
  page: number;
  pageSize: number;
}

export interface ReviewStats {
  total: number;
  visible: number;
  hidden: number;
  averageStars: number;
  fiveStars: number;
  fourStars: number;
  threeStars: number;
  twoStars: number;
  oneStar: number;
}

export interface ReviewOverview {
  total: number;
  pending: number;
  thisWeek: number;
  averageStars: number;
  healthScore: number;
  positiveRate: number;
  withReply: number;
  fiveStars: number;
  fourStars: number;
  threeStars: number;
  twoStars: number;
  oneStar: number;
  visible: number;
  hidden: number;
  byCategory: Record<string, number>;
  byStatus: Record<string, number>;
}

export interface ReviewTrendPoint {
  month: string;
  averageStars: number;
  count: number;
}

export interface AdminReviewQuery {
  page?: number;
  pageSize?: number;
  status?: ReviewStatus | '';
  category?: ReviewCategory | '';
  minStars?: number | null;
  maxStars?: number | null;
  sortBy?: string;
}

// ── Usage Intelligence ─────────────────────────────────────────────────────

export interface UsageIntelligenceOverview {
  totalAiRequests: number;
  flaggedRequests: number;
  usersNeedingReview: number;
  warningsSent: number;
}

export interface UsageAttentionUser {
  userId: string;
  userName: string;
  userEmail: string;
  profileImageUrl?: string | null;
  totalRequests: number;
  flaggedRequests: number;
  warningsSent: number;
  lastActivity?: string | null;
}

export interface UsageFlaggedRequest {
  id: string;
  requestType: string;
  userRequest: string;
  aiResponse: string;
  classification: string;
  reason: string;
  createdAt: string;
}

export interface UsageAdminAction {
  id: string;
  actionType: string;
  adminName: string;
  notes?: string | null;
  createdAt: string;
}

export interface UsageUserDetail {
  userId: string;
  userName: string;
  userEmail: string;
  profileImageUrl?: string | null;
  totalRequests: number;
  flaggedRequests: number;
  warningsSent: number;
  lastActivity?: string | null;
  isAiRestricted: boolean;
  isRestricted: boolean;
  isSuspended: boolean;
  flaggedItems: UsageFlaggedRequest[];
  actionHistory: UsageAdminAction[];
}

export interface TakeUsageActionRequest {
  actionType: string;
  notes?: string | null;
}

// ── Audit Logs ─────────────────────────────────────────────────────────────

export type AuditSeverity = 'Info' | 'Success' | 'Warning' | 'Critical';
export type AuditModule   = 'Users' | 'Reviews' | 'AI Operations' | 'Settings' | 'System';

export interface AuditChange {
  field: string;
  before: string | null;
  after:  string | null;
}

export interface AuditLogEntry {
  id:          string;
  timestamp:   string;
  actorName:   string;
  actorEmail:  string;
  module:      AuditModule;
  action:      string;
  target:      string;
  severity:    AuditSeverity;
  description: string;
  changes:     AuditChange[];
}

export interface AuditLogSummary {
  total:         number;
  today:         number;
  critical:      number;
  adminActions:  number;
  infoCount:     number;
  successCount:  number;
  warningCount:  number;
  criticalCount: number;
}

export interface AuditLogQuery {
  search?:   string;
  module?:   string;
  severity?: string;
  dateFrom?: string;
  dateTo?:   string;
  page?:     number;
  pageSize?: number;
}
